xrootd/protocol/protocol.go

```diff
@@ -102,3 +106,97 @@ const (
 	// DataServer indicates whether this is a data server.
 	DataServer
 )
+
+// EntryStat holds the entry name and the entry stat information.
+type EntryStat struct {
+	EntryName     string // EntryName is the name of entry.
+	HasStatInfo   bool   // HasStatInfo indicates if the following stat information is valid.
+	ID            int64  // ID is the OS-dependent identifier assigned to this entry.
+	EntrySize     int64  // EntrySize is the decimal size of the entry.
+	IsExecutable  bool   // IsExecutable indicates that entry is either an executable file or a searchable directory.
+	IsDirectory   bool   // IsDirectory indicates that entry is a directory.
+	IsOther       bool   // IsOther indicates that entry is neither a file nor a directory.
+	IsOffline     bool   // IsOffline indicates that the file is not online (i. e., on disk).
+	IsReadable    bool   // IsReadable indicates that read access to that entry is allowed.
+	IsWritable    bool   // IsWritable indicates that write access to that entry is allowed.
+	IsPoscPending bool   // IsPoscPending indicates that the file was created with kXR_posc and has not yet been successfully closed.
+	Mtime         int64  // Mtime is the last modification time in Unix time units.
+}
+
+func (es EntryStat) Name() string {
+	return es.EntryName
+}
+
+func (es EntryStat) Size() int64 {
+	return es.EntrySize
+}
+
+func (es EntryStat) ModTime() time.Time {
+	return time.Unix(es.Mtime, 0)
+}
+
+func (es EntryStat) IsDir() bool {
+	return es.IsDirectory
+}
+
+func (es EntryStat) Sys() interface{} {
+	return nil
+}
+
+func (es EntryStat) Mode() os.FileMode {
+	var mode os.FileMode
+	if es.IsDir() {
+		mode |= os.ModeDir
+	}
+	if es.IsWritable {
+		mode |= 0222
+	}
+	if es.IsReadable {
+		mode |= 0444
+	}
+	return mode
+}
+
+// NewEntryStat creates a new EntryStat according to the name of the entry
+// and statinfo following XRootD protocol specification (see page 114).
+func NewEntryStat(name, statinfo string) (EntryStat, error) {
+	stats := strings.Split(statinfo, " ")
+
+	if len(stats) < 4 {
+		return EntryStat{}, errors.Errorf("xrootd: statinfo \"%s\" doesn't have enough fields, expected format is: \"id size flags modtime\"", statinfo)
+	}
+
+	id, err := strconv.Atoi(stats[0])
+	if err != nil {
+		return EntryStat{}, err
+	}
+	size, err := strconv.Atoi(stats[1])
+	if err != nil {
+		return EntryStat{}, err
+	}
+	flags, err := strconv.Atoi(stats[2])
+	if err != nil {
+		return EntryStat{}, err
+	}
+	mtime, err := strconv.Atoi(stats[3])
+	if err != nil {
+		return EntryStat{}, err
+	}
+
+	result := EntryStat{
+		EntryName:     name,
+		HasStatInfo:   true,
+		ID:            int64(id),
+		EntrySize:     int64(size),
+		Mtime:         int64(mtime),
+		IsExecutable:  flags&1 != 0,
```

Comment:
perhaps store `flags` as an `EntryStat` field and convert all these `IsXyz` fields into methods on `EntryStat`?